### ISSUE TITLE
Why so many connect info wrapper?

### ISSUE THREAD

**Issue by winterland1989** [naushadh/persistent#1]
Now i got the time to give a review on this, the first question is that it seems there're more one `Connectinfo` wrappers, i can understand the motivation, but does it necessary?

**Comment by naushadh** [comment]
I suppose you're referring to `MySQLConnectInfo` (Public version, part of `Database.Persist.MySQL`) and `MySQLConf`.

I'll refer to the `persistent-mysql-haskell` version of `MySQLConnectInfo` with a (P) suffix and `mysql-haskell` version with an (M) suffix here on out.

- `MySQLConf` was ported over from `persistent-mysql` before its constructor and named property functions were hidden for encapsulation. I'm not sure how it can be removed or cleanly merged with `MySQLConnectInfo` (P).
- `MySQLConnectInfo` (P) is new, and [it initially started of as a `newtype`](https://github.com/naushadh/persistent/commit/9f4f4065f823a871eb85e59a038b5f850152be2d#diff-3c6aeb99ebb433bed56cfa31255978fdR893) wrapper for `MySQLConnectInfo` (M). Even if we ignore the benefits on encapsulation (more on that later) I had documented in README, I _had_ to use it to [support TLS](https://github.com/naushadh/persistent/commit/6217f2a7400fc7cd4c7e8978012751a6bceef46c#diff-3c6aeb99ebb433bed56cfa31255978fdR910-913) without drastically altering the other APIs (to accept a Maybe ClientInfo as an additional param). Then there was the issue of should someone else ask for SSL support, I'll have to create even more APIs or make breaking changes to existing ones. With `MySQLConnectInfo` (P) I can easily support TLS, and potentially also SSL side by side. (i.e, setting one unsets the other).

I'd like to hear more on how you envision the wrappers be changed/merged/removed.

P.S. Forward compatibility sidenote: Way back before `persisent-mysql-haskell` was a thing, I tried to update [`persistent-mysql` to use `mysql-haskell` _and_ keep the APIs the same](https://github.com/naushadh/persistent/commit/69ff5e7d02fd971e15504f6a2cc2af105557439e), but I fell short on supporting some features that were too `libmysqlclient` specific. I wished `persistent-mysql` didn't expose as many `libmysqlclient` features and had an encapsulated type so the transition to `mysql-haskell` would have been seamless. I hope having `MySQLConnectInfo` (P) would allow for someone else in the future attempting to change something big internally, especially important if `persistent-mysql-haskell` reaches high adoption directly or indirectly (being merged upstream).
